storagenode/pieces/blobcache.go

```diff
@@ -0,0 +1,272 @@
+			blobs.cache.total,
+		)
+	}
+
+	blobs.mu.Lock()
+	blobs.cache.total = estimatedTotals
+	blobs.cache.totalBySatellite = estimatedTotalsBySatellite
+	blobs.mu.Unlock()
+	return nil
+}
+
+func estimate(newSpaceUsedTotal, spaceUsedWhenIterationStarted, spaceUsedWhenIterationEnded int64) int64 {
+	if newSpaceUsedTotal == spaceUsedWhenIterationEnded {
+		return newSpaceUsedTotal
+	}
+	spaceUsedDeltaDuringIteration := spaceUsedWhenIterationEnded - spaceUsedWhenIterationStarted
+	return newSpaceUsedTotal + (spaceUsedDeltaDuringIteration / 2)
+}
+
+// Delete gets the size of the piece that is going to be deleted then deletes it and
+// updates the space used cache accordingly
+func (blobs *BlobsUsageCache) Delete(ctx context.Context, blobRef storage.BlobRef) error {
+	blobInfo, err := blobs.Stat(ctx, blobRef)
+	if err != nil {
+		return err
+	}
+	// calling with nil for store since we don't need it to
+	// get content size
+	pieceAccess, err := newStoredPieceAccess(nil, blobInfo)
+	if err != nil {
+		return err
+	}
+	pieceContentSize, err := pieceAccess.ContentSize(ctx)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	err = blobs.Blobs.Delete(ctx, blobRef)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	satelliteID := storj.NodeID{}
+	satelliteID.Scan(blobRef.Namespace)
+	blobs.update(ctx, satelliteID, pieceContentSize)
+	return nil
+}
+
+// Create returns a blobWriter that knows which namespace/satellite its writing the piece to
+// and also has access to the space used cache to update when finished writing the new piece
+func (blobs *BlobsUsageCache) Create(ctx context.Context, ref storage.BlobRef, size int64) (_ storage.BlobWriter, err error) {
+	blobWriter, err := blobs.Blobs.Create(ctx, ref, size)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+	return &blobCacheWriter{
+		BlobWriter: blobWriter,
+		usageCache: blobs,
+		namespace:  string(ref.Namespace),
+	}, nil
+}
+
+// TestCreateV0 creates a new V0 blob that can be written. This is only appropriate in test situations.
+func (blobs *BlobsUsageCache) TestCreateV0(ctx context.Context, ref storage.BlobRef) (_ storage.BlobWriter, err error) {
+	fStore := blobs.Blobs.(interface {
+		TestCreateV0(ctx context.Context, ref storage.BlobRef) (_ storage.BlobWriter, err error)
+	})
+	return fStore.TestCreateV0(ctx, ref)
+}
+
+type blobCacheWriter struct {
+	storage.BlobWriter
+	usageCache *BlobsUsageCache
+	namespace  string
+}
+
+// Commit updates the cache with the size of the new piece that was just
+// created then it calls the blobWriter to complete the upload.
+func (blob *blobCacheWriter) Commit(ctx context.Context) error {
+	// get the size written we commit that way this
+	// value will only include the piece content size and not
+	// the header bytes
+	size, err := blob.BlobWriter.Size()
+	if err != nil {
+		return Error.Wrap(err)
+	}
+	satelliteID := storj.NodeID{}
+	satelliteID.Scan(blob.namespace)
+	blob.usageCache.update(ctx, satelliteID, size)
```

Comment:
since the usage cache update can't fail, shouldn't we do it after the commit?

Reply:
i was debating doing if after or before commit. all the commit does is write the header bytes, the piece content is already written. that is why i updated before

Reply:
sure, but if it’s after, you don’t have to know what commit does to agree the behavior is right. you’re sure that update only happens on correctly committed pieces.

Reply:
good point. i will move the update to after the commit